src/platform/nrfconnect/FactoryDataProvider.h

```diff
@@ -102,10 +102,74 @@ struct ExternalFlashFactoryData
     uint8_t mFactoryDataBuffer[PM_FACTORY_DATA_SIZE];
 };
 
+class FactoryDataProviderProxy : public chip::Credentials::DeviceAttestationCredentialsProvider,
+                                 public CommissionableDataProvider,
+                                 public DeviceInstanceInfoProvider
+{
+public:
+    virtual CHIP_ERROR Init() { return CHIP_NO_ERROR; };
+
+    // ===== Members functions that are platform-specific
+    virtual CHIP_ERROR GetEnableKey(MutableByteSpan & enableKey) { return CHIP_NO_ERROR; }
+
+    /**
+     * @brief Get the user data in CBOR format as MutableByteSpan
+     *
+     * @param userData MutableByteSpan object to obtain all user data in CBOR format
+     * @returns
+     * CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND if factory data does not contain user field, or the value cannot be read out.
+     * CHIP_ERROR_BUFFER_TOO_SMALL if provided MutableByteSpan is too small
+     */
+    virtual CHIP_ERROR GetUserData(MutableByteSpan & userData) { return CHIP_NO_ERROR; }
+
+    /**
+     * @brief Try to find user data key and return its value
+     *
+     * @param userKey A key name to be found
+     * @param buf Buffer to store value of found key
+     * @param len Length of the buffer. This value will be updated to the actual value if the key is read.
+     * @returns
+     * CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND if factory data does not contain user field, or the value cannot be read out.
+     * CHIP_ERROR_BUFFER_TOO_SMALL if provided buffer length is too small
+     */
+    virtual CHIP_ERROR GetUserKey(const char * userKey, void * buf, size_t & len) { return CHIP_NO_ERROR; }
+
+protected:
+    // ===== Members functions that implement the DeviceAttestationCredentialsProvider
+    CHIP_ERROR GetCertificationDeclaration(MutableByteSpan & outBuffer) override { return CHIP_NO_ERROR; };
+    CHIP_ERROR GetFirmwareInformation(MutableByteSpan & out_firmware_info_buffer) override { return CHIP_NO_ERROR; };
```

Comment:
Should we return in each of the methods "CHIP_NO_ERROR"? Shouldn't be CHIP_ERROR_NOT_IMPLEMENTED returned?

Reply:
IMHO we shouldn't even provide any default implementations. We still want to force the concrete implementations to provide their own implementations, so it's the best not to provide any defaults so that the compiler may step in if something is missing.

Reply:
I agree, changed the interface to purely virtual.